ui: Upgrade React version and react-dependent libraries

koorosh opened the pull request:
- Has to merged after: https://github.com/cockroachdb/cockroach/pull/43685
- Depends on https://github.com/cockroachdb/yarn-vendored/pull/13

Current PR has affected a lot of files but most of changes are almost the same
and do not affect business logic.

With current changes, react is upgraded to latest version (16.12.0) and
following libraries are upgraded as well: react-redux, redux, redux-saga.
These upgrades of major versions has required to do some code adjustments
to properly migrate to newest versions.
- redux. Provide valid type arguments for Dispatch type.
- redux-saga. Changed import and usage of 'delay' function.
- react-redux. `connect` function failed to correctly resolve returned type
for `mapDispatchToProps` argument. As result it is provided as a function
instead of object argument. This hack allows implicitly resolve types.
- `CachedDataReducer` class is extended with one more optional type argument
to specify the list of allowed `actionNamespace` literals instead of `string`.
It was necessary to have strictly defined CachedDataReducer interface.
- `normalizeConnectedComponent` function is used to do simple mapping from
react's ExoticComponent to valid react component. This resolves issue of
providing ConnectedComponents (which are ExoticComponent's, not regular react
components) to Route component which has strict type validation of component types.

koorosh committed:
ui: Upgrade React version and react-dependent libraries

With current changes, react is upgraded to latest version (16.12.0) and
following libraries are upgraded as well: react-redux, redux, redux-saga.
These upgrades of major versions has required to do some code adjustments
to properly migrate to newest versions.
- redux. Provide valid type arguments for Dispatch type.
- redux-saga. Changed import and usage of 'delay' function.
- react-redux. `connect` function failed to correctly resolve returned type
for `mapDispatchToProps` argument. As result it is provided as a function
instead of object argument. This hack allows implicitly resolve types.
- `CachedDataReducer` class is extended with one more optional type argument
to specify the list of allowed `actionNamespace` literals instead of `string`.
It was necessary to have strictly defined CachedDataReducer interface.
- `normalizeConnectedComponent` function is used to do simple mapping from
react's ExoticComponent to valid react component. This resolves issue of
providing ConnectedComponents (which are ExoticComponent's, not regular react
components) to Route component which has strict type validation of component types.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43846)
<!-- Reviewable:end -->

Koorosh commented on pkg/ui/ccl/src/views/shared/containers/licenseSwap/index.tsx:
```diff
@@ -53,11 +53,11 @@ export default function swapByLicense<TProps>(
   const cclName = getComponentName(CCLComponent);
 
-  class LicenseSwap extends React.Component<TProps & OwnProps> {
+  class LicenseSwap extends React.Component<TProps & OwnProps & any> {
```
Following hack with **&any** type is required to make changes related to types only and do not rewrite component logic.

koorosh committed:
ui: Add license header for new file

Release note: None

koorosh committed:
ui: Resolve conflicts after rebasing against master

- tests command is disabled for now because current state doesn't
contain all fixes to run tests without failure.
- Decommissioned Node History page was added so code has to be adjusted.

Release note: None

koorosh committed:
ui: Fix binding actions to react component with actionCreator

Redux Actions for some components (ClusterOverview, NodesLogs, Statements page, and
Node Maps) were not properly binded. As result component were not provided with data,
and weren't rendered.
To fix this, actions are wrapped with 'bindActionCreators' helper.

Release note: None

koorosh committed:
ui: Fix "Maximum call stack size" issue with react-helmet library

- Issue: nfl/react-helmet#373
- Workaround: nfl/react-helmet#373 (comment)

Release note: None

dhartunian reviewed (changes requested):
Thanks for this huge undertaking @Koorosh! My only major concern is whether the restructuring the `mapDispatchToProps` is necessary. See my comments for the approach I looked at.

koorosh committed:
ui: Refactor mapping of actions to components

- get rid of unnecessary wrapping action creators with
bindActionCreators function
- define correct return type for cachedDataReducer.refresh action

Release note: None

koorosh committed:
ui: Align heading for network page with text

Release note: None

koorosh committed:
ui: Simplify mapping Action dispatchers to component properties

Remove useless function wrapper for action creator object

Release note: None

Koorosh commented on pkg/ui/src/views/cluster/containers/events/index.tsx:
```diff
@@ -178,9 +176,9 @@ const eventBoxConnected = connect(
     };
   },
-  {
+  () => ({
```
Agree. Removed unnecessary function wrappers across all project

Koorosh commented on pkg/ui/src/redux/cachedDataReducer.ts:
```diff
@@ -180,7 +180,7 @@ export class CachedDataReducer<TRequest, TResponseMessage> {
    */
   refresh = <S>(req?: TRequest, stateAccessor = (state: any, _req: TRequest) => state.cachedData[this.actionNamespace]) => {
-    return (dispatch: Dispatch<S>, getState: () => any) => {
+    return (dispatch: Dispatch<Action, TResponseMessage>, getState: () => S) => {
```
Awesome, it makes code much cleaner! Thanks!

koorosh committed:
ui: Update ref to yarn-vendor sub-module

Release note: None

dhartunian reviewed (approved):
👍 LGTM. Thanks for the updates!

dhartunian commented:
bors r+
